daemon/commit.go

```diff
@@ -149,6 +149,11 @@ func (daemon *Daemon) Commit(name string, c *backend.ContainerCommitConfig) (str
 		defer daemon.containerUnpause(container)
 	}
 
+	// Ensure environ substitutions in commit use the container's image config
+	if len(c.Config.Env) == 0 {
+		c.Config.Env = make([]string, len(container.Config.Env))
+		copy(c.Config.Env, container.Config.Env)
```

Comment:
I'm not sure why `c.Config` even exists. It seems like it should be using `container.Config` as the base, but it's been this way for a while. 

This fix looks right when used with the CLI, because the CLI never sends a `c.Config` in the request body, but there could be API consumers sending a request body. This is a great example of why testing using the CLI is a bad idea.

If someone is sending a Config in the request body I guess it's on them to use the correct Config.

Reply:
I think the _actual_ semantics I want is to replace this if statement [here](https://github.com/moby/moby/blob/c307e0ce49d63d1b1f479560955dca3956f005ec/api/server/router/image/image_routes.go#L49-L51) to something like:

```
if c == nil {
    c = copyRunConfig(GetContainer(cname))
}
```
- `copyRunConfig` being [from here](https://github.com/moby/moby/blob/c307e0ce49d63d1b1f479560955dca3956f005ec/builder/dockerfile/internals.go#L434)
- `GetContainer` being [from here](https://github.com/moby/moby/blob/c307e0ce49d63d1b1f479560955dca3956f005ec/daemon/container.go#L34)

but the current encapsulation doesn't allow me to do this (perhaps intentionally?)

Reply:
You shouldn't need copy because `GetContainer()` should hopefully give you a copy already.

`GetContainer()` could be exposed in the `Backend` interface. That should probably work

Reply:
from what I can tell, it gives me the live `Container` and not a copy :(

Where would I look to extend the `Backend` interface?

Reply:
https://github.com/moby/moby/blob/c307e0ce49d63d1b1f479560955dca3956f005ec/api/server/router/image/backend.go

The methods exposed by backend all return `api/types` not internal types, so I guess you'll need to use `Daemon.ContainerInspectCurrent()` instead of `GetContainer()`, but that should still give you the `container.Config`, and that one would be a copy.